tgui-next: Initial PR, speedy new Air Alarm interface.

stylemistake opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request

Read Issue #46246 

Some interfaces (one of them is Air Alarm UI) were rewritten in a completely new UI framework based on Inferno. New UI is very fast.

The rest of not implemented yet interfaces are delegated to an old tgui by dynamically loading old tgui js/css. Dynamic loading might slow down old tgui interfaces.

## Why It's Good For The Game

**[100% LAG FREE SCRUBBERS](https://cdn.discordapp.com/attachments/326831214667235328/626653924975443968/2019-09-26_08-37-48.mp4)**

## Changelog
:cl:
add: New React-based TGUI, which is INCREDIBLY FAST.
add: New Acclimator interface (@actioninja).
add: New AI Airlock interface (@actioninja).
add: New Air Alarm interface (@stylemistake).
add: New Airlock Electronics interface (@actioninja).
add: New APC interface (@actioninja).
add: New Atmos Alert Console interface (@actioninja).
add: New Atmos Control Console interface (@actioninja).
add: New Atmos Filter interface (@actioninja).
add: New Atmos Mixer interface (@actioninja).
add: New Atmos Pump interface (@actioninja).
add: New Borg Panel interface (@actioninja).
add: New Brig Timer interface (@actioninja).
add: New Bluespace Artillery interface (@actioninja).
add: New Gas Canister interface (@willox).
add: New Cargo interface (@actioninja, @stylemistake).
add: New Cellular Emporium interface for changelings (@actioninja).
add: New CentCom Pod Launcher interface (@actioninja).
add: New Chem Dispenser interface (@stylemistake).
add: New Crayon interface (@actioninja).
add: New Disposal Unit interface (@actioninja).
add: New Thermomachine interface (@willox).
add: New Wires (hacking) interface (@actioninja).
tweak: Old TGUI interfaces are dynamically loaded if new TGUI interface was not found. This might increase load times while we are undergoing the transition.
tweak: Scrubber list is more compact, using chemical formula notation for common gases.
tweak: Slightly more native-looking title bar in new TGUI.
tweak: Window scrollbars are now themed to match TGUI color scheme.
fix: New TGUI hopefully fixes all previous window dragging and resizing issues.
code: Very robust UI development pipeline, with hot module replacement and live log collection.
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

stylemistake committed:
Initial tgui-next commit

stylemistake committed:
Improve dev server, slight changes to layout.

stylemistake committed:
Add more components, fix dragging, finalize scrubber list

stylemistake committed:
Complete an air alarm interface.

Qustinnus commented:
I dislike the idea of introducing another UI system without removing the old, how will we prevent people from ignoring the old templates forever and then prevent those templates from causing maintenance issues?

Qustinnus commented:
Besides that epic pr keep it up gamer

stylemistake commented:
> I dislike the idea of introducing another UI system without removing the old, how will we prevent people from ignoring the old templates forever and then prevent those templates from causing maintenance issues?

If we don't keep both systems in place, we might just never transition to the new system, because the amount of work necessary to move all interfaces is pretty huge.

Qustinnus commented:
okay but like we have 3 formats right now, I see it as a possible problem. I think straight out banning tgui changes is a good idea

JJRcop commented on tgui-next/packages/functional/index.js:
```diff
@@ -0,0 +1,47 @@
+/**
+ * @file
+ * @copyright 2018 Aleksej Komarov
+ * @license GPL-2.0-or-later
```
Highlighting license for maintainers to review

stylemistake replied:
@JJRcop it is compatible with AGPL v3, because of the GPL-2.0-**or-later**. I can remove licensing too, implicitly upgrading it to AGPL v3, but affero's extra clauses are pointless for javascript.

MrStonedOne replied:
>but affero's extra clauses are pointless for javascript.

No they aren't! .min files are considered object files and AGPL forces anybody using them to provide the original version.

JJRcop commented:
Would the autodoc system work with these JS files?

81Denton commented:
@Qustinnus If this proves to be a viable full replacement, we can just change our CONTRIBUTING.md

willox commented:
I don't know if it is intentional but this lacks the behaviour from TGUI where the buttons dim a little bit with UI_UPDATE (the orange eye in the top-left corner) and grey out fully when disabled (the red eye.)

You can reproduce both by just standing different distances from the air alarm.

actioninja commented:
I'm not a big fan of requiring yarn on top of node, but playing around with this a bit the hotloader and live logs are incredible.

stylemistake commented:
> Would the autodoc system work with these JS files?

@JJRcop yes.

> I don't know if it is intentional but this lacks the behaviour from TGUI where the buttons dim a little bit with UI_UPDATE or UI_DISABLED (the orange/red eye in the top-left corner). You can reproduce both by just standing different distances from the air alarm.

Not implemented yet. I think I'll bring a next patch of fixes today, fixing this in the process too.

---

Also, a daily reminder...

## pls test on IE8 who has wine

Thanks, this is for you @Cyprex in particular, because you got it working, and I would waste a lot of time setting it up. Ping me if you need a pre-built DM binary.

actioninja committed:
acclimator + ai airlock ui

stylemistake committed:
Multiple improvements based on feedback

- LabeledList now has a "buttons" prop, where you can put buttons.
- Improved Box and Flex components
- Whole UI is now dimmable if non-interactive
- Basic documentation of components (Box and Flex so far).
- Icon and Box now accept a "color" prop.
- Routing improved in "Layout" component, you can now specify whether an interface is scrollable.

stylemistake committed:
Less harsh dimming

stylemistake committed:
Redux, Toasts

- Fixed inconsistent backend updates with Redux.
- Added Toasts which are currently unused, but are functional.

actioninja committed:
better ai door interact ui

actioninja committed:
final ai airlock interface

stylemistake committed:
Add a progress bar component, implement resizing

stylemistake commented:
I think now we have a full feature parity with old TGUI.

- Dimming was addressed.
- Added a ProgressBar component, with CSS animations.
- Added a Toast component, without CSS animations yet, but functional.
- Window resizing, on all left to bottom sides of a window.

Remaining fixes:

- FPS drop if you load complex interfaces and do a lot of rerenders. Same problem existed on an old TGUI.
- ???

[Progress demo 1](https://cdn.discordapp.com/attachments/326831214667235328/627474339234709514/2019-09-28_14-58-02.mp4)
[Progress demo 2](https://cdn.discordapp.com/attachments/326831214667235328/627476837458640896/2019-09-28_15-07-18.mp4)

@actioninja was busy porting other interfaces (they will land in a different PR):

![image](https://media.discordapp.net/attachments/326831214667235328/627467920645160966/unknown.png)

stylemistake committed:
Fix a zero in title bar

willox commented on tgui-next/packages/tgui/interfaces/AirAlarm.js:
```diff
@@ -0,0 +1,481 @@
+  const { config, data } = state;
+  const { ref } = config;
+  const entries = data.environment_data || [];
+  const dangerMap = {
```
The localStatusText values here are in the wrong order.

stylemistake committed:
Add a linter to keep shit tidy, fix danger level mapping, add some more docs

PKPenguin321 commented:
>Would the autodoc system work with these JS files?

It seems to already have documentation in `tgui-next/README.md`, though I haven't checked to see how comprehensive it is. I would hope it contains basic development instructions that developers can use as a reference for making new UIs, which was notoriously difficult with tgui.

stylemistake commented:
I added "next" to the folder so that when we are done, we kebab the old one, and this one becomes simply "tgui".

stylemistake committed:
Fix issues with code, enforce hard 120 line length cap, automerge binaries

stylemistake committed:
Implement hot module reloading

stylemistake committed:
Fix progress bar, add color support

stylemistake committed:
Fix ProgressBar baseline alignment issues

stylemistake committed:
Remove unwanted padding on the bottom of the LabeledList.

stylemistake committed:
Component improvements

- Fix baseline issues with Button and ProgressBar components
- Box how inherits props from Box
- Atomic colors and Button colors are now auto-generated, all range of colors supported

stylemistake committed:
Chem Dispenser UI, animated numbers, more style fixes

stylemistake committed:
Add an IE8 polyfill

SpaceManiac commented:
I have IE8 set up, will test when able

spookydonut commented on tgui-next/yarn.lock:
```diff
@@ -0,0 +1,6231 @@
+# THIS IS AN AUTOGENERATED FILE. DO NOT EDIT THIS FILE DIRECTLY.
```
should this file be in .gitignore or what

stylemistake replied:
No, because it's necessary for the package manager to behave deterministically, aka it precisely locks all dependency versions so that you always get predictable installs on any system, and it's not really updating that much.

MrStonedOne replied:
The true answer here is that YARN is being stupid and abusing the `.lock` extension in a way that creates confusion. To anybody unfamiliar with YARN or other equally stupid package managers, .lock is normally an extension used for temp per-session purposes, and not for long term storage.

stylemistake committed:
Intermediate state of fixing IE8

Cyprex commented:
Throwing up netcat on port 30001 gave me an IE 7 useragent for byond webviews, standalone IE(iexplore.exe) in the wineprefix gives IE 8. "nc -l -p 30001" gave
>User-Agent: Mozilla/4.0 (compatible; MSIE 7.0; Windows NT 5.1; Trident/4.0; byond_5.0; .NET CLR 2.0.50727; .NET CLR 3.0.30729; .NET CLR 3.5.30729; .NET4.0C; .NET4.0E)

for byond webviews. Standalone IE in the same wineprefix produces

>User-Agent: Mozilla/4.0 (compatible; MSIE 8.0; Windows NT 5.1; Trident/4.0; byond_5.0; .NET CLR 2.0.50727; .NET CLR 3.0.30729; .NET CLR 3.5.30729; .NET4.0C; .NET4.0E)

The requests from byond were triggered by replacing tgui.bundle.js with the following three-line snippet:
>var t = new XMLHttpRequest();
>t.open("GET", 'http://127.0.0.1:30001', 0);
t>.send();

SpaceManiac commented:
@Cyprex A particular meta tag needs to be near the start of `<head>` for BYOND's embedded webview to use the most recent rendering engine it can:

```html
<meta http-equiv="X-UA-Compatible" content="IE=edge" />
````

This is true to get IE11 features on an IE11-supporting system too.

stylemistake committed:
Lots of shimming and general misery

stylemistake commented:
Alright, I got very good news. New UI is now officially usable on IE8.

Bad news: Old UI is not loading due to unknown issues. So far [this have been the main suspect](https://github.com/ractivejs/ractive/issues/2343), although it's only a guess, because just about everything in IE8 is a blackbox. Can't debug anything at all.

I've added 2 new polyfills (console stubs and CSS object model) in addition to 3 already existing, and heavily reworked the build pipeline to account for oddities of ActiveX™ JScript and Webpack/Babel abusing `Object.defineProperty` pretty heavily if not keeping them under control. Polyfills are now located externally in public folder as separate files, which are loaded into the browser through use of conditional IE tags.

> A particular meta tag needs to be near the start of <head> for BYOND's embedded webview to use the most recent rendering engine it can

It still reports its User Agent as IE 7, no matter how you position this tag, and it does not affect rendering at all. We tried to detect visual differences between compatibility and no compatibility, and we found none.

optimumtact commented:
given the lack of movement in a long time on tgui im happy to approve this pr provisionally, even if I think the buildchain suffers from similar problems that might lead down a similar path of never getting any updates in.

stylemistake commented:
> Old UI is not loading due to unknown issues.

Small update: Old UI is not loading for me even on public servers.

stylemistake committed:
Fully working old TGUI chainloader for IE8, more pipeline improvements

stylemistake commented:
Alright folks, we got IE8 completely sorted.

> Old UI is not loading due to unknown issues.

This was fixed by creating a separate page `tgui-fallback.html`, which only has scripts necessary to run the old tgui. One problem is that `ref` will be lost in the process because this is a plain HTML file. Solution is extremely retarded, but it works nicely: we redirect the webview to `tgui-fallback.html?ref=${ref}`, and in a new page we reconstruct the template by doing this:

```js
var ref = location.search.replace('?ref=', '');
document.getElementById('data').setAttribute('data-ref', ref);
```

We do not need to reconstruct the state, because backend will send a fresh state anyway after calling `tgui:initialize`.

This also by proxy solves #46847, here is an image of it working:
![image](https://cdn.discordapp.com/attachments/628449393435213852/629121287394164747/unknown.png)

For Linux folks: `tgui-dev-tools` now fully work on Linux, with live compilation, cache reloader and live log collection (via basic HTTP requests). It tries to search for your BYOND cache folder in `~/.wine` and `~/Games/byond` wine prefixes. So you can enjoy all benefits of development on Windows, and get that UI done and tested quicker.

On this note, I call for testmerging, and no more features will be added to this PR.

stylemistake committed:
Support title-less Sections

MrStonedOne commented on tgui-next/.editorconfig:
```diff
@@ -0,0 +1,13 @@
+root = true
+
+[*]
+indent_style = space
```
Please fix

stylemistake replied:
Fix what exactly

MrStonedOne replied:
This project uses tabs.

You are adding tgui-next to this project, so it too should use tabs.

MrStonedOne commented on tgui-next/.eslintrc.yml:
```diff
@@ -0,0 +1,118 @@
+  array-bracket-newline: [error, consistent]
+  array-bracket-spacing: [error, never]
+  block-spacing: [error, always]
+  brace-style: [error, stroustrup, { allowSingleLine: false }]
```
Please set to `1tbs`

stylemistake replied:
perhaps we can just simply relax some of these rules, including this one

MrStonedOne commented on tgui-next/.eslintrc.yml:
```diff
@@ -0,0 +1,118 @@
+  func-style: [error, expression]
+  ## This rule does not honor a newline on opening paren.
+  # function-paren-newline: [error, never]
+  indent: [error, 2, {
```
Please change the second index from `2` to `tab`

stylemistake replied:
If i am not wrong, tools like npm and yarn use spaces regardless of tab usage in file or editorconfig setting, and spaces-as-tabs are used almost in every js project out there. If you're using snippets of javascript from the web, you will be forced to convert it to tabs. Old tgui also uses spaces instead of tabs, that's why I tried to reflect this in editor settings.

Linter is also capable of fixing styling issues automatically by running it with an extra `--fix` flag.

spookydonut replied:
There are some files in this repo still using space indentation and it's shit.

MrStonedOne commented:
Please read through https://github.com/tgstation/tgstation/blob/master/.github/CONTRIBUTING.md and ensure your linter is not enforcing different coding styles from the rest of the project (where applicable).

You will not be the only one modifying these JS files. You **MUST** make this accessible to somebody who has never worked in js, never worked with managed web frameworks, and has only coded in DM, as that describes an uncomfortably large percent of our contributor base.

The first step to doing that is ensuring they don't get burned by conflicting style enforcement, so tgui-next's linter and editor config should have settings designed to be as close as possible to the rest of the project's style guidelines.

OR!

Maybe add the linter in a seperate PR so this PR doesn't get bogged down with arguments over that.

stylemistake commented:
We can replace Yarn with Lerna:

https://github.com/lerna/lerna

This will allow us to continue to use a multi-package setup. This is still a separate tool, but it can be safely installed locally with npm, and we can add a post-install hook to bootstrap all packages. From a developer's perspective, you only have to write `npm install`.

Installation procedure will take a bit longer than yarn, though, and package-lock.json is still badly implemented in NPM, which might cause long term issues with dependencies randomly updating, etc.

stylemistake committed:
Delete Yarn, use Lerna for workspace management

stylemistake commented:
I have rebased the branch, and replaced Yarn with Lerna. So essentially, you no longer need to install Yarn.

optimumtact commented:
they should just learn to code

stylemistake commented:
Testmerge is not rolling out well, because a lot of people still get an old tgui. All javascript and css assets are properly updated in the cache, the problem is that [an old HTML template](https://github.com/stylemistake/tgstation/blob/tgui-next/code/controllers/subsystem/tgui.dm#L13-L14) is served, which is a part of an rsc file.

This might be somehow related to this: http://www.byond.com/forum/post/103188

stylemistake committed:
Improve maths and animated numbers

stylemistake committed:
Fix loss of focus for scrollable content, activate buttons with keyboard

stylemistake committed:
Attempt to bust the cache, grab focus on keyboard presses too

stylemistake committed:
Fix hot module replacement by manually pulling replaced modules.

optimumtact commented:
@MrStonedOne is your CDN fuxxored?

MrStonedOne commented:
The CDN only has resources, not assets, assets are handled by the asset cache and byond's `browse_rsc` system, and that doesn't cache byond the session. To be exact, the asset cache framework will cache per-connection, and DS will cache per-process-session, but caches based off of BYONDcrc, not filename.

It shouldn't be that, i might try to reproduce if somebody can give me a step to reproduce and check, mainly like what interface to open and how to know old vs new layout.

actioninja committed:
backend logging

 un-nuke line endings

changes without insane line endings

actioninja committed:
helper procs + href support

actioninja committed:
slight optimization

actioninja committed:
compiles

stylemistake committed:
Redux, Hotkeys, Kitchen Sink (for UI testing), Tabs component

stylemistake committed:
Push logs to backend, small kitchen sink changes, tab fixes

stylemistake committed:
Update component reference in README

stylemistake committed:
Small fixes

stylemistake commented:
@MrStonedOne
> if somebody can give me a step to reproduce
There is no clear way to reproduce it, some of us got a new UI, some of us didn't. So far it's a total mistery.

> ...and check, mainly like what interface to open and how to know old vs new layout.
New Air Alarm UI can be easily identified by a glowing red close button when you hover it, as on [this screenshot](https://user-images.githubusercontent.com/1516236/65765905-1ea11b80-e132-11e9-8d13-2514c8a0c118.png).

stylemistake committed:
Next bunch of IE8 fixes, PureComponent optimization for Inferno

stylemistake committed:
Delete old tgui html

stylemistake committed:
Log the event of loading the old tgui

Cyprex commented:
Tested most of the UIs migrated recently on a local server. Everything seems to work in IE 8.

Testmerge has been updated on Terry by oranges.

stylemistake committed:
Enable tgui logging on server by default

stylemistake commented:
@MrStonedOne nothing from the code side of things is helping with this cached html issue. I tried loading html in runtime, even deleted the old html file, and I still receive an old tgui on Terry (which is currently on commit 1d8859a). I really think it has something to do with deploy code. Please help me fix this because this is critically important and we can't merge it in this state.

optimumtact commented:
@MrStonedOne there definiteyl seems to be some issue where they're being served the old html and cleaning up the entire local cache isn't helping, it's either an issue with tgs not doing a clean compile, or the cloudflare cdn is caching the base html and serving a different thing to different players.

stylemistake committed:
Final solution

stylemistake committed:
Extract routes into a separate file, fix ChemDispenser bug

- Chem dispenser was needlessly disabling transfer amount buttons

stylemistake committed:
Disable baseline alignment for vertical tabs

stylemistake committed:
Fix tabs and box units

- Tab content was not taking full page width
- Box can now accepts strings as units, which will be treated as raw style values.

stylemistake committed:
Fix tgui on Terry

stylemistake committed:
Fix sending all logs with an "error" flag

JJRcop commented:
How is the documentation as of now, does it adequately cover important functions and syntax of tgui2? How well are the internals documented for future contributors?

stylemistake commented:
README has a pretty good documentation of UI components and what you can do with them.

Almost all helper functions in the "common" package have docstrings, and they are extensively used througout the code, so you learn about them by example.

I do admit that state management and Redux thing needs a little bit of explanation, because it's not very obvious. You can make some **really advanced** pieces of functionality with it. What we currently have is support for hotkeys and toast notifications. This is all largely undocumented and a work in progress.

If you a complete newbie however, you can ignore that it even exists. For all you care, it just provides your components with the state from the server (you get it via the `state` prop), and that's about it.

As of the internals, the build pipeline, I don't know if it's really that important to document. I mean, code is self explanatory for the most part (if you are willing to read it), there is very little magic happening behind the scenes.

stylemistake commented:
> syntax of tgui2

README refers to React documentation, which explains pretty much everything, and if you are new to React, you are **required** to read it.

stylemistake committed:
Some macro UI component changes and improvements

- Refer to README.md diff to see what's new

stylemistake committed:
Tooltip component

stylemistake committed:
Add support for themes, nanotrasen logo is back

stylemistake committed:
Clockwork theme hotfix

stylemistake commented:
Full feature parity has been reached, IE8 works, additional failsafe was added to keep old tgui working.

Ready for the merge.

actioninja committed:
Lots of stuff backported from actioninja's repo

stylemistake committed:
Slight adjustment to logo positioning

stylemistake committed:
Actually proper solution for logo positioning

stylemistake committed:
Fix color maps

Anonmare commented:
I don't think this is respecting APC modes as if the pwoer is low and I try to force equipment to ON, the APC will force it to OFF

sohpeach committed:
tgui-next thermomachine interface

sohpeach committed:
tgui-next canister interface

stylemistake committed:
Add icon_class_name proc to asset cache

stylemistake committed:
Cleanup oopsies

stylemistake committed:
Cargo UI adjustments

Cyprex commented:
d89f9de broke IE 8 support: Exception thrown and not caught (http://127.0.0.1:51789/tmp349/tgui.bundle.js line 1)

actioninja commented:
It's most likely not due to any of my changes, but commits were squashes a ton anyways. That commit also isn't entirely my code.

Cyprex commented:
That commit came up during bisection, could you link me to some source commits that I could check to further narrow down the source?

actioninja commented:
you could try pulling `tgui-next-ui-dev` off of my branch and see if it's still broken. If it isn't, then it's more likely to be tweaks made by style himself.

Cyprex commented:
fef0d06bbca45a6bbde9b26b1c4691d3833f8629 , "ui dev work" on that branch seems to have introducted this issue.

stylemistake committed:
Nuke lodash

stylemistake committed:
Minor fixes and cleanup

- Remove local Inferno package in favor of upstream Inferno
- Fix "initial" prop on AnimatedNumber
- Force fancy=0 for IE8
- Click-through dimmer

stylemistake committed:
Add a bat file for dummies

TheChosenEvilOne commented on tgui-next/packages/tgui/interfaces/CentcomPodlauncher.js:
```diff
@@ -0,0 +1,313 @@
+              content="Specific Target"
+              selected={data.effectBluespace}
+              tooltip="Gives the supplypod an advanced Bluespace Recyling Device. After opening, the supplypod will be warped directly to the surface of a nearby NT-designated trash planet (/r/ss13)."
+              onClick={() => act(ref, 'effectBluespace')} />
```
hmmm, this does not look right.

stylemistake replied:
@actioninja

actioninja committed:
podlauncher interface fix

stylemistake committed:
Update README, assert code dominance

stylemistake committed:
Clarify usage of event handlers in Inferno

stylemistake committed:
Document LabeledList.Divider

JJRcop commented:
If maintainers believe the README needs more they will let you know. I've removed the label since you've created quite the amount of documentation, and now it's up to review.

stylemistake committed:
Fukken grammar

actioninja committed:
fixes cellular emporium readapt button not working

stylemistake commented:
Some nice-to-haves to implement later:
- Warn about unused `onClick` handler on buttons, because that is a potential bug, but I don't know how to set it up in such a way that would be annoying for devs when prototyping. Perhaps emit the warning on production builds, to be logged in round logs.
- Warn about lowercase `onclick` handlers, which is also a pretty common source of bugs.
- Emit errors when inserting illegal children elements, like `<br />` into `<LabeledList>` or inserting something else than a `<Tabs.Tab>` into `<Tabs>`.

willox commented on tgui-next/packages/tgui/interfaces/AtmosControlConsole.js:
```diff
@@ -0,0 +1,86 @@
+                <Button
+                  icon="pencil-alt"
+                  content={round(data.inputRate) + ' L/s'}
+                  onClick={() => act(ref, 'input')} />
```
should be `onClick={() => act(ref, 'rate')} />`

actioninja committed:
fixes incorrect action on button in atmos control interface

actioninja committed:
remove unneeded data from airlock electronics ui

stylemistake committed:
Set +x mode on bin/tgui

stylemistake committed:
Fix filename cases

- They were untracked by git on windows

stylemistake committed:
Ignore package lock, make batch more robust

- package-lock.json is very random and unreliable as fuck. Yarn was better.

stylemistake committed:
Build tgui-next in travis

stylemistake committed:
bruh

- fixes tgui error reporting

AnturK commented on code/modules/client/asset_cache.dm:
```diff
@@ -325,6 +325,13 @@ GLOBAL_LIST_EMPTY(asset_datums)
 	var/size_id = sprite[SPR_SIZE]
 	return {"<span class="[name][size_id] [sprite_name]"></span>"}
```
should use the new helper here for consistency.

stylemistake replied:
You mean we can't leave `icon_class_name` helper to be used for future interfaces? Just saying, Inferno can't render html tags like that.

AnturK reviewed (changes requested):
DM side stuff, i will be looking over js shortly.

actioninja committed:
logging cleanup + always log

Kraseo commented:
alien multitools do not show what wire does which when you hold it in hand with this interface, would be neat if it could be fixed

stylemistake commented on tgui-next/packages/tgui-dev-server/link/client.js:
```diff
@@ -0,0 +1,108 @@
+            })
+            // .then(modules => {
+            //   sendLogEntry(null, 'outdated modules', modules);
+            // })
```
This code was used for debugging hot module replacement.

Cyprex commented:
I think borg wires are affected as well, I didn't get a legend with blueprints in the offhand
